core/src/main/java/org/opensearch/sql/expression/function/PPLFuncImpTable.java

```diff
@@ -0,0 +1,345 @@
+    List<RelDataType> ANY_TYPE_1 = List.of(ANY_TYPE);
+
+    RexNode apply(RexBuilder builder, RexNode arg1);
+
+    @Override
+    default RexNode apply(RexBuilder builder, RexNode... args) {
+      if (args.length != 1) {
+        throw new IllegalArgumentException("This function requires exactly 1 arguments");
+      }
+      return apply(builder, args[0]);
+    }
+
+    @Override
+    default List<RelDataType> getParams() {
+      return ANY_TYPE_1;
+    }
+  }
+
+  public interface FunctionImp2 extends FunctionImp {
+    List<RelDataType> ANY_TYPE_2 = List.of(ANY_TYPE, ANY_TYPE);
+
+    RexNode apply(RexBuilder builder, RexNode arg1, RexNode arg2);
+
+    @Override
+    default RexNode apply(RexBuilder builder, RexNode... args) {
+      if (args.length != 2) {
+        throw new IllegalArgumentException("This function requires exactly 2 arguments");
+      }
+      return apply(builder, args[0], args[1]);
+    }
+
+    @Override
+    default List<RelDataType> getParams() {
+      return ANY_TYPE_2;
+    }
+  }
+
+  /** The singleton instance. */
+  public static final PPLFuncImpTable INSTANCE;
+
+  static {
+    final Builder builder = new Builder();
+    builder.populate();
+    INSTANCE = new PPLFuncImpTable(builder);
+  }
+
+  private final ImmutableMap<BuiltinFunctionName, PairList<CalciteFuncSignature, FunctionImp>> map;
+
+  private PPLFuncImpTable(Builder builder) {
+    final ImmutableMap.Builder<BuiltinFunctionName, PairList<CalciteFuncSignature, FunctionImp>>
+        mapBuilder = ImmutableMap.builder();
+    builder.map.forEach((k, v) -> mapBuilder.put(k, v.immutable()));
+    this.map = ImmutableMap.copyOf(mapBuilder.build());
+  }
+
+  public @Nullable FunctionImp get(final BuiltinFunctionName functionName) {
+    final PairList<CalciteFuncSignature, FunctionImp> implementList =
+        requireNonNull(map.get(functionName));
+    if (implementList.isEmpty()) {
+      throw new NullPointerException();
+    }
+    return implementList.getFirst().getValue();
+  }
+
+  public @Nullable FunctionImp resolveSafe(final String functionName, List<RexNode> args) {
+    try {
+      return resolve(functionName, args);
+    } catch (IllegalStateException e) {
+      return null;
+    }
+  }
+
+  public FunctionImp resolve(final String functionName, List<RexNode> args) {
+    Optional<BuiltinFunctionName> funcNameOpt = BuiltinFunctionName.of(functionName);
+    if (funcNameOpt.isEmpty()) {
+      throw new IllegalArgumentException(String.format("Unsupported function: %s", functionName));
+    }
+    BuiltinFunctionName.of(functionName);
+    return resolve(funcNameOpt.get(), args);
+  }
+
+  public FunctionImp resolve(final BuiltinFunctionName functionName, List<RexNode> args) {
+    final PairList<CalciteFuncSignature, FunctionImp> implementList = map.get(functionName);
+    if (implementList == null || implementList.isEmpty()) {
+      throw new IllegalStateException(String.format("Cannot resolve function: %s", functionName));
```

Comment:
why `IllegalStateException`? do you mean `IllegalArgumentException`? Or `SemanticCheckException`